### PR TITLE
osd: Weighted Random Sampling for dynamic perf stats

### PR DESCRIPTION
**Pull request by trociny**
Signed-off-by: Mykola Golub <mgolub@suse.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by trociny** [comment]
@tchaikov It is not related (it is a race in that unit test) but thank you for reporting. I created a tracker ticket for it [1].

[1] https://tracker.ceph.com/issues/37672

File: src/osd/DynamicPerfStats.h [hunk 1]
```diff
@@ -204,30 +204,39 @@ class DynamicPerfStats {
...
+        uint64_t wsum = 0;
+        for (size_t i = 0; i < limit.max_count; i++) {
+          wsum += it_counters->second[index].first;
+          counter_iterators.push_back(it_counters++);
+        }
+        for (; it_counters != counters.end(); it_counters++) {
+          wsum += it_counters->second[index].first;
+          double p = wsum > 0 ?
+            1.0 * it_counters->second[index].first / wsum : 0;
+          double j = 1.0 * rand() / RAND_MAX;
```
Comment (dillaman) [inline src/osd/DynamicPerfStats.h]: Nit: can we replace the use of ```rand``` with the C++ ```<random>``` uniform distribution number generator helpers?

**Comment by trociny** [comment]
@dillaman updated

File: src/osd/DynamicPerfStats.h [hunk 2]
```diff
@@ -204,30 +205,38 @@ class DynamicPerfStats {
...
+
+        Iterator it_counters = counters.begin();
+        uint64_t wsum = 0;
+        for (size_t i = 0; i < limit.max_count; i++) {
+          wsum += it_counters->second[index].first;
+          counter_iterators.push_back(it_counters++);
+        }
+        for (; it_counters != counters.end(); it_counters++) {
+          wsum += it_counters->second[index].first;
+          if (ceph::util::generate_random_number(0, wsum) <=
```
Comment (dillaman) [inline src/osd/DynamicPerfStats.h]: You should use a class-level instance of ```random_number_generator``` and use its functor operator to generate each number. Using the method directly will just initialize a new rng engine for each function call.

Comment (trociny) [reply src/osd/DynamicPerfStats.h]: @dillaman I thought `ceph::util::generate_random_number` will reuse `thread_local rng_engine` form  "include/random.h"? No?

Comment (dillaman) [reply src/osd/DynamicPerfStats.h]: Missed that implementation detail -- thx.

**Comment by dillaman** [comment]
retest this please

**Pull request merged**